// @esri/hub-sites/dist/esm/layout/convert-section.js
import { cloneObject, getProp } from "@esri/hub-common";
import { convertRow } from "./convert-row";
import { extractAssets } from "./extract-assets";
/**
 * Convert a section, collecting assets along the way...
 * @param {ISection} section the section to templatize
 */
export function convertSection(section) {
    var _a;
    // if the section has a background image, and it has a url, we should
    // add that to the asset hash so it can be downloaded and added to the template item
    // and also cook some unique asset name so we can inject a placeholder
    var _b = section.rows.reduce(toTemplatizedRows, {
        assets: [],
        rows: []
    }), rows = _b.rows, assets = _b.assets;
    var result = {
        section: cloneObject(section),
        assets: assets
    };
    result.section.rows = rows;
    if (sectionHasBackgroundFile(section)) {
        (_a = result.assets).push.apply(_a, extractAssets(section.style.background));
    }
    return result;
}
function toTemplatizedRows(acc, row) {
    var _a;
    var _b = convertRow(row), assets = _b.assets, cards = _b.cards;
    (_a = acc.assets).push.apply(_a, assets);
    acc.rows.push({ cards: cards });
    return acc;
}
function sectionHasBackgroundFile(clonedSection) {
    return getProp(clonedSection, "style.background.fileSrc");
}
