// @esri/hub-sites/dist/node/update-app-redirect-uris.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.updateAppRedirectUris = void 0;
const arcgis_rest_request_1 = require("@esri/arcgis-rest-request");
const hub_common_1 = require("@esri/hub-common");
/**
 * Update the redirect uri's that are valid for an existing app that's registered
 * for oAuth.
 * @param {string} clientId Client Id of the existing app to be updated
 * @param {Array} redirectUris Array of valid redirect uris for the app
 * @param {IRequestOptions} requestOptions
 */
function updateAppRedirectUris(clientId, redirectUris, requestOptions) {
    const url = `${hub_common_1.getPortalApiUrl(requestOptions)}/oauth2/apps/${clientId}/update`;
    const options = {
        method: "POST",
        authentication: requestOptions.authentication,
        params: {
            client_id: clientId,
            redirect_uris: JSON.stringify(redirectUris)
        }
    };
    return arcgis_rest_request_1.request(url, options);
}
exports.updateAppRedirectUris = updateAppRedirectUris;
