// @esri/hub-sites/dist/esm/drafts/save-draft.js
import { objectToJsonBlob, getProp, buildDraft, } from "@esri/hub-common";
import { _includeListFromItemType } from "./_include-list-from-item-type";
import { deleteDraft } from "./delete-draft";
import { addItemResource } from "@esri/arcgis-rest-portal";
/**
 * Given a site or page model, saves a draft
 *
 * NOTE - replaces current draft if exists
 * @param {*} siteOrPageModel
 * @param {*} hubRequestOptions
 */
export function saveDraft(siteOrPageModel, hubRequestOptions) {
    var includeList = _includeListFromItemType(siteOrPageModel.item);
    var draft = buildDraft(siteOrPageModel, includeList);
    var draftBlob = objectToJsonBlob(draft);
    var draftName = "draft-" + Date.now() + ".json";
    var itemId = getProp(siteOrPageModel, "item.id");
    return deleteDraft(siteOrPageModel, hubRequestOptions)
        .then(function (_) {
        return addItemResource({
            id: itemId,
            owner: getProp(siteOrPageModel, "item.owner"),
            resource: draftBlob,
            name: draftName,
            private: false,
            portal: hubRequestOptions.portal,
            authentication: hubRequestOptions.authentication,
        });
    })
        .then(function (_) { return draft; });
}
