// @esri/hub-sites/dist/node/layout/convert-card.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertCard = void 0;
const hub_common_1 = require("@esri/hub-common");
const converters = {
    "event-list-card": convertEventListCard,
    "follow-initiative-card": convertFollowCard,
    "items/gallery-card": convertItemGalleryCard,
    "image-card": convertImageCard,
    "jumbotron-card": convertImageCard
};
/**
 * Convert a card to a templatized version of itself
 * @param {ICard} card the card to templatize
 */
function convertCard(card) {
    const clone = hub_common_1.cloneObject(card);
    const converter = converters[clone.component.name];
    if (converter) {
        return converter(clone);
    }
    return {
        card: clone,
        assets: []
    };
}
exports.convertCard = convertCard;
function convertEventListCard(card) {
    card.component.settings.initiativeIds = ["{{initiative.item.id}}"];
    return { card, assets: [] };
}
function convertFollowCard(card) {
    card.component.settings.initiativeId = "{{initiative.item.id}}";
    return { card, assets: [] };
}
function convertImageCard(card) {
    const result = {
        card,
        assets: []
    };
    if (hub_common_1.getProp(card, "component.settings.fileSrc")) {
        result.assets.push(card.component.settings.fileSrc);
    }
    if (hub_common_1.getProp(card, "component.settings.cropSrc")) {
        result.assets.push(card.component.settings.cropSrc);
    }
    return result;
}
function convertItemGalleryCard(card) {
    const settings = card.component.settings;
    const version = hub_common_1.getProp(settings, "version");
    if (hub_common_1.getProp(settings, "groups") && version < 4) {
        settings.groups = [
            {
                title: "{{solution.title}}",
                id: "{{teams.contentGroupId}}"
            }
        ];
    }
    if (hub_common_1.getProp(settings, "query.groups")) {
        if (version >= 4) {
            settings.query.groups = ["{{teams.contentGroupId}}"];
        }
        else {
            settings.query.groups = [
                {
                    title: "{{solution.title}}",
                    id: "{{teams.contentGroupId}}"
                }
            ];
        }
    }
    if (hub_common_1.getProp(settings, "query.orgId")) {
        settings.query.orgId = "{{organization.id}}";
    }
    if (hub_common_1.getProp(settings, "orgId")) {
        settings.orgId = "{{organization.id}}";
    }
    if (settings.siteId) {
        settings.siteId = "{{appid}}";
    }
    return { card, assets: [] };
}
