// @esri/hub-sites/dist/node/layout/convert-section.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertSection = void 0;
const hub_common_1 = require("@esri/hub-common");
const convert_row_1 = require("./convert-row");
const extract_assets_1 = require("./extract-assets");
/**
 * Convert a section, collecting assets along the way...
 * @param {ISection} section the section to templatize
 */
function convertSection(section) {
    // if the section has a background image, and it has a url, we should
    // add that to the asset hash so it can be downloaded and added to the template item
    // and also cook some unique asset name so we can inject a placeholder
    const { rows, assets } = section.rows.reduce(toTemplatizedRows, {
        assets: [],
        rows: []
    });
    const result = {
        section: hub_common_1.cloneObject(section),
        assets
    };
    result.section.rows = rows;
    if (sectionHasBackgroundFile(section)) {
        result.assets.push(...extract_assets_1.extractAssets(section.style.background));
    }
    return result;
}
exports.convertSection = convertSection;
function toTemplatizedRows(acc, row) {
    const { assets, cards } = convert_row_1.convertRow(row);
    acc.assets.push(...assets);
    acc.rows.push({ cards });
    return acc;
}
function sectionHasBackgroundFile(clonedSection) {
    return hub_common_1.getProp(clonedSection, "style.background.fileSrc");
}
