// @esri/hub-sites/dist/node/domains/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var hub_common_1 = require("@esri/hub-common");
/* istanbul ignore next */
Object.defineProperty(exports, "addDomain", { enumerable: true, get: function () { return hub_common_1.addDomain; } });
/* istanbul ignore next */
Object.defineProperty(exports, "domainExistsPortal", { enumerable: true, get: function () { return hub_common_1.domainExistsPortal; } });
/* istanbul ignore next */
Object.defineProperty(exports, "domainExists", { enumerable: true, get: function () { return hub_common_1.domainExists; } });
/* istanbul ignore next */
Object.defineProperty(exports, "ensureUniqueDomainName", { enumerable: true, get: function () { return hub_common_1.ensureUniqueDomainName; } });
/* istanbul ignore next */
Object.defineProperty(exports, "getDomainsForSite", { enumerable: true, get: function () { return hub_common_1.getDomainsForSite; } });
/* istanbul ignore next */
Object.defineProperty(exports, "getUniqueDomainNamePortal", { enumerable: true, get: function () { return hub_common_1.getUniqueDomainNamePortal; } });
/* istanbul ignore next */
Object.defineProperty(exports, "getUniqueDomainName", { enumerable: true, get: function () { return hub_common_1.getUniqueDomainName; } });
/* istanbul ignore next */
Object.defineProperty(exports, "isDomainForLegacySite", { enumerable: true, get: function () { return hub_common_1.isDomainForLegacySite; } });
/* istanbul ignore next */
Object.defineProperty(exports, "isDomainUsedElsewhere", { enumerable: true, get: function () { return hub_common_1.isDomainUsedElsewhere; } });
/* istanbul ignore next */
Object.defineProperty(exports, "isValidDomain", { enumerable: true, get: function () { return hub_common_1.isValidDomain; } });
/* istanbul ignore next */
Object.defineProperty(exports, "lookupDomain", { enumerable: true, get: function () { return hub_common_1.lookupDomain; } });
/* istanbul ignore next */
Object.defineProperty(exports, "removeDomain", { enumerable: true, get: function () { return hub_common_1.removeDomain; } });
/* istanbul ignore next */
Object.defineProperty(exports, "updateDomain", { enumerable: true, get: function () { return hub_common_1.updateDomain; } });
