// @esri/hub-sites/dist/node/drafts/save-draft.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.saveDraft = void 0;
const hub_common_1 = require("@esri/hub-common");
const _include_list_from_item_type_1 = require("./_include-list-from-item-type");
const delete_draft_1 = require("./delete-draft");
const arcgis_rest_portal_1 = require("@esri/arcgis-rest-portal");
/**
 * Given a site or page model, saves a draft
 *
 * NOTE - replaces current draft if exists
 * @param {*} siteOrPageModel
 * @param {*} hubRequestOptions
 */
function saveDraft(siteOrPageModel, hubRequestOptions) {
    const includeList = _include_list_from_item_type_1._includeListFromItemType(siteOrPageModel.item);
    const draft = hub_common_1.buildDraft(siteOrPageModel, includeList);
    const draftBlob = hub_common_1.objectToJsonBlob(draft);
    const draftName = `draft-${Date.now()}.json`;
    const itemId = hub_common_1.getProp(siteOrPageModel, "item.id");
    return delete_draft_1.deleteDraft(siteOrPageModel, hubRequestOptions)
        .then((_) => arcgis_rest_portal_1.addItemResource({
        id: itemId,
        owner: hub_common_1.getProp(siteOrPageModel, "item.owner"),
        resource: draftBlob,
        name: draftName,
        private: false,
        portal: hubRequestOptions.portal,
        authentication: hubRequestOptions.authentication,
    }))
        .then((_) => draft);
}
exports.saveDraft = saveDraft;
